content/md/en/docs/tutorials/work-with-pallets/use-macros-in-a-custom-pallet.md

```diff
@@ -262,54 +303,63 @@ To implement this logic in the proof-of-existence pallet:
-        let current_block = <frame_system::Pallet<T>>::block_number();
-
-        // Store the claim with the sender and block number.
-        Claims::<T>::insert(&claim, (&sender, current_block));
-
-        // Emit an event that the claim was created.
-        Self::deposit_event(Event::ClaimCreated { who: sender, claim });
-
-        Ok(())
-      }
-
-      #[pallet::weight(0)]
-      pub fn revoke_claim(origin: OriginFor<T>, claim: T::Hash) -> DispatchResult {
-        // Check that the extrinsic was signed and get the signer.
-        // This function will return an error if the extrinsic is not signed.
-        let sender = ensure_signed(origin)?;
-
-        // Get owner of the claim, if none return an error.
-        let (owner, _) = Claims::<T>::get(&claim).ok_or(Error::<T>::NoSuchClaim)?;
-
-        // Verify that sender of the current call is the claim owner.
-        ensure!(sender == owner, Error::<T>::NotClaimOwner);
-
-        // Remove claim from storage.
-        Claims::<T>::remove(&claim);
-
-        // Emit an event that the claim was erased.
-        Self::deposit_event(Event::ClaimRevoked { who: sender, claim });
-        Ok(())
-      }
-    }
-    ```
+   ```rust
+   // Dispatchable functions allow users to interact with the pallet and invoke state changes.
+   // These functions materialize as "extrinsics", which are often compared to transactions.
+   // Dispatchable functions must be annotated with a weight and must return a DispatchResult.
+   #[pallet::call]
+   impl<T: Config> Pallet<T> {
+   	#[pallet::weight(1_000)]
+   	pub fn create_claim(
+   		origin: OriginFor<T>,
+   		proof: Vec<u8>,
+   	) -> DispatchResult {
+   		// Check that the extrinsic was signed and get the signer.
+   		// This function will return an error if the extrinsic is not signed.
+   		let sender = ensure_signed(origin)?;
+
+   		// Verify that the specified proof has not already been claimed.
+   		ensure!(!Proofs::<T>::contains_key(&proof), Error::<T>::ProofAlreadyClaimed);
+
+   		// Get the block number from the FRAME System pallet.
+   		let current_block = <frame_system::Pallet<T>>::block_number();
+
+   		// Store the proof with the sender and block number.
+   		Proofs::<T>::insert(&proof, (&sender, current_block));
+
+   		// Emit an event that the claim was created.
+   		Self::deposit_event(Event::ClaimCreated(sender, proof));
+
+   		Ok(())
+   	}
+
+   	#[pallet::weight(10_000)]
+   	pub fn revoke_claim(
+   		origin: OriginFor<T>,
+   		proof: Vec<u8>,
+   	) -> DispatchResult {
+   		// Check that the extrinsic was signed and get the signer.
+   		// This function will return an error if the extrinsic is not signed.
+   		let sender = ensure_signed(origin)?;
+
+   		// Verify that the specified proof has been claimed.
+   		ensure!(Proofs::<T>::contains_key(&proof), Error::<T>::NoSuchProof);
+
+   		// Get owner of the claim.
+			let (owner, _) = Proofs::<T>::get(&proof).expect("This should not fail; qed");
```

Comment:
WTF!

NO!

Never panic in the runtime! You can simply return an error here

This is also an awful proof cc @sacha-l